### PR TITLE
EEP: Maven-style Plugins

### PR DESCRIPTION
**Pull request by dmikurube**
This is an "retrospective" EEP (Embulk Enhancement Proposal) for the Maven-style Plugins.

### CODE REVIEW

**Comment by dmikurube** [comment]
Jfyi, updated with user configurations for Maven-style plugins.

**Comment by dmikurube** [comment]
Added another explanation about the main class and the manifest.

File: docs/eeps/draft-maven-style-plugins.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    * JRuby takes a long time to initialize itself.
+    * Making JRuby optional is another effort to be explained in another EEP. This EEP would compose a part of the effort.
+* Bundler is mandatory to select a specific version of the same RubyGems-style plugin if multiple versions are installed.
+    * Users sometimes wanted to select a specific version from the same installation set.
+    * Reconfiguring Bunlder needs time, and sometimes reinstallation.
+* A RubyGems-style plugin has to include all the dependency JAR files packaged.
+    * The size of a RubyGems-style plugin package tended to be large.
+    * Many Embulk plugins are expected to depend on the same Java libraries. They would be duplicated on the network for downloading, and on the file system.
+    * License matters are more complicated in a RubyGems-style plugin because it includes redistributions.
+
```
Comment (hiroyuki-sato) [inline docs/eeps/draft-maven-style-plugins.md:31]: The Maven-style plugin allows the same plugin name. The RubyGem-style plugin doesn't allow it. How about adding that description as motivation?

**Review by hiroyuki-sato — COMMENTED** [review]
(I'm still reading Main class and Manifest part)

Which plugin will be loaded if a user installs the same name, RubyGems-style and Maven-style style plugins?
(How about adding a description about plugin load ordering?)

**Comment by dmikurube** [comment]
Ah, thanks for your good points! I'll add explaining them.

**Comment by dmikurube** [comment]
@hiroyuki-sato Addressed your comments at ee71259692152ca1b7752bf1ea9e5643c0bf55f8

File: docs/eeps/draft-maven-style-plugins.md [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+These JAR files are the "classpath" to be loaded for the Maven-style plugin. Embulk can load these classes now, but Embulk still needs to identify the "main class" as an Embulk plugin, which implements Embulk SPI's `*Plugin` interface, such as `org.embulk.spi.InputPlugin`, `org.embulk.spi.FilterPlugin`, and else.
+
+The main class is to be declared in `META-INF/MANIFEST.MF`, [the manifest file](https://docs.oracle.com/javase/tutorial/deployment/jar/manifestindex.html) of the plugin JAR file. The manifest file of an Embulk plugin's JAR file must include the following attributes, `Embulk-Plugin-Spi-Version`, `Embulk-Plugin-Category`, `Embulk-Plugin-Main-Class`, and `Embulk-Plugin-Type`.
+
+* `Embulk-Plugin-Spi-Version` must be `0` as of Embulk v0.11 or earlier.
+* `Embulk-Plugin-Category` is the category of the plugin, which is one of `input`, `decoder`, `parser`, `filter`, `formatter`, `encoder`, `output`, or `guess`.
```
Comment (hiroyuki-sato) [inline docs/eeps/draft-maven-style-plugins.md:208]: (I'm not familiar `pom.xml` format.)

A few days ago, @dmikurube and I talked about `embulk-parser-csv` packaging. [Twitter](https://twitter.com/dmikurube/status/1666743761295245313?s=20)
We will plan a parser and a guess package into a single jar file.
In this case, Does this attribute care about that?
(I'm wondering **one of** `parser`, .. or `guess` parts).

In those case, Does A Plugin developer creates two manifests?

Comment (dmikurube) [reply docs/eeps/draft-maven-style-plugins.md:208]: @dmikurube This part in this EEP explains only about each JAR package == each Maven artifact. `pom.xml` is per Maven artifact.

On the other hand, what we talked on `embulk-parser-csv` was not about "packaging" nor each artifact. I was talking only about the Git source code repository, not about the release package.

It intended about having two (or more) plugin source trees in the same Git repository like : 

```
embulk-parser-guess.git
    + embulk-parser-csv/
    | + build.gradle
    | + src/
    | + ...
    + embulk-guess-csv/
      + build.gradle
      + src/
      + ...
```

[`embulk-input-jdbc`](https://github.com/embulk/embulk-input-jdbc) is a similar example doing the same.

Comment (hiroyuki-sato) [reply docs/eeps/draft-maven-style-plugins.md:208]: Ah! I got it. Thanks

File: docs/eeps/draft-maven-style-plugins.md [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+* `Embulk-Plugin-Main-Class` is the "main class" of the Embulk plugin, which implements Embulk SPI's `*Plugin` interface.
+* `Embulk-Plugin-Type` is the type of the Embulk plugin, such as `file`, `s3`, `csv`, `gzip`, or else.
+
+See the following example of a manifest file.
```
Comment (hiroyuki-sato) [inline docs/eeps/draft-maven-style-plugins.md:212]: How about writing a descritpon about Non `Embulk-` prefix attributes? (I a bit confused about `Implementation-Title: embulk-input-file` )
ex) Non `Embulk` prefix attribute description are described in the manifest file.

**Review by hiroyuki-sato — APPROVED** [review]
LGTM 👍 (I commented the one minor thing)

File: docs/eeps/draft-maven-style-plugins.md (hunk 3 above)
Comment (dmikurube) [reply docs/eeps/draft-maven-style-plugins.md:212]: They're Java's standard ones, and explained in a Java page linked from the URL above.

https://docs.oracle.com/javase/tutorial/deployment/jar/packageman.html

Comment (dmikurube) [reply docs/eeps/draft-maven-style-plugins.md]: Added a quick note:

> Note that `Manifest-Version`, `Implementation-Title`, and `Implementation-Version` are Java standard ones. See ["Working with Manifest Files: The Basics"](https://docs.oracle.com/javase/tutorial/deployment/jar/manifestindex.html) for more details.

**Comment by dmikurube** [comment]
Thanks! Let's merge it as EEP-5 as Final.

**Comment by hiroyuki-sato** [comment]
LGTM 👍

**Pull request merged**